book/html.md

```diff
@@ -5,408 +5,559 @@ prev: text
+    "link", "meta", "title", "style", "script",
+]
+```
+
+That tells you whether to insert `<head>` or `<body>`:
+
+``` {.python indent=8}
+elif open_tags == ["html"] and tag not in ["head", "body", "/html"]:
+    if tag in HEAD_TAGS:
+        implicit = "head"
+    else:
+        implicit = "body"
+    currently_open.append(ElementNode(implicit))
 ```
 
-The new `layout_close` will be pretty similar (unsetting the flags),
-while `layout_text` will do the line wrapping and terminal space
-computations from the last chapter.
+If you see an element that's not supposed to go in the `<head>`, you
+need to implicitly close the `<head>` section:
 
-Finally, don\'t forget to update `show` to get its display list from
-inside `state`.
+``` {.python indent=8}
+elif open_tags == ["html", "head"] and tag not in ["/head"] + HEAD_TAGS:
+    node = currently_open.pop()
+    currently_open[-1].children.append(node)
+    currently_open.append(ElementNode("body"))
+```
+
+In all other cases, implicit tags aren't needed, and we can exit the
+loop:
+
+``` {.python indent=8}
+else:
+    break
+```
+
+We haven't yet handled the implicit close tags `</body>` and
+`</html>`. Usually, leaving them out will mean the `parse` function
+reaches the end of its loop without closing all open tags, and would
+thus return nothing. Let's make it return *something* instead:
 
 ``` {.python}
-layout(nodes)
-display_list = state.display_list
+def parse(tokens):
+    # ...
+    while currently_open:
+        node = currently_open.pop()
+        if not currently_open: return node
+        currently_open[-1].children.append(node)
 ```
 
+These rules for malformed HTML may seem arbitrary, and they are: they
+evolved over years of trying to guess what people "meant" when they
+wrote that HTML, and are now codified in the [HTML parsing
+standard](html5-parsing).
+
+::: {.further}
+HTML parsers also have an [algorithm][adoption] to handle misnested
+elements, plus a [list of active formatting
+elements][html5-formatting-list] to handle formatting like
+`<b>b<i>bi</b>i</i>`.
+:::
+
+[html5-formatting-list]: https://html.spec.whatwg.org/multipage/parsing.html#the-list-of-active-formatting-elements
+[adoption]: https://html.spec.whatwg.org/multipage/parsing.html#adoption-agency-algorithm
 
 Summary
 =======
 
-In this chapter, we taught our browser to understand HTML as a
-structured tree, not just a flat list of tokens, and we\'ve updated
-layout to be a recursive tree traversal instead of a linear pass through
-the document. We\'ve also made the browser much more robust to malformed
-HTML. While these changes don\'t have much impact yet, this new,
-structured understanding of HTML sets us up to implement a layout
-engine in the next chapter.
+This chapter taught our browser that HTML is a tree, not just a flat
+list of tokens:
+
+- A parser now transforms HTML tokens to a tree
+- Layout operates recursively on the tree
+- The lexer recognizes and handles attributes on elements
```

Comment:
```suggestion
- Lexer code to recognize and handle attributes on elements
```

(did you introduce what a "lexer" is?)

Reply:
Good point! I call the function `lex` but don't think I ever called it a lexer.